File: Miscellaneous/Golang/234. Palindrome Linked List.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func main() {

	list := &ListNode{1, nil}
	list.Next = &ListNode{2, nil}
	list.Next.Next = &ListNode{2, nil}
	list.Next.Next.Next = &ListNode{1, nil}
	result := isPalindrome(list)
	fmt.Println(result)
}

func isPalindrome(head *ListNode) bool {
	temp := head
	var stack []int

	for temp != nil {
		stack = append(stack, temp.Val)
		temp = temp.Next
	}

	for head != nil {
		val := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		if val != head.Val {
			return false
		}
		head = head.Next
	}

	return true
}

File: Miscellaneous/Python/234. Palindrome Linked List.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def isPalindrome(self, head):
        stack = []
        temp = head
        
        while temp is not None:
            stack.append(temp.val)
            temp = temp.next
        
        while head is not None:
            val = stack.pop()
            if head.val is not val:
                return False
            head = head.next
       
        
        return True
        
obj = Solution()
list = ListNode(1)
list.next = ListNode(2)
list.next.next = ListNode(2)
list.next.next.next = ListNode(1)
ans = obj.isPalindrome(list)
print(ans)